x/staking/spec/02_state_transitions.md

```diff
@@ -92,6 +100,8 @@ occur when the unbonding delegation queue element matures:
 - remove the entry from the `UnbondingDelegation` object
 - transfer the tokens from the `NotBondedPool` `ModuleAccount` to the delegator `Account`
 
+Thoughts: Complete unbonding should take effect on epoching but transition from bonded to unbonding could take effect instantly.
```

Comment:
I think this can work either between epochs or at the end of epochs.  Is there a reason for why you prefer at the end of an epoch?